app/controllers/key_pairs_controller.rb

```diff
@@ -37,18 +37,50 @@ def edit
     @key_pair = KeyPair.find(params[:id])
   end
 
+  # Given a File object to upload, save it on the file system with
+  # association to the user_name given.
+  # Admins are able to save keys under different user names then
+  # their own.
+  # Creates the KEY_STORAGE directory if it does not yet exist
+  def upload_key_file(_file, user_name)
+    Dir.mkdir(KEY_STORAGE) unless File.exists?(KEY_STORAGE)
+    uploaded_io = _file
+
+    # todo: check current user to see if they have permission to
+    # use a different user_name then their current one (admins)
+
+    File.open(Rails.root.join(KEY_STORAGE, user_name + '.pub'), 'wb') do |f|
+      f.write(uploaded_io.read)
+    end
+
+  end
+
   # POST /key_pairs
   # POST /key_pairs.json
   def create
-    @key_pair = KeyPair.new(key_pair_params)
+
+    # Upload the file
+    upload_key_file(key_pair_params[:file], @current_user.user_name)
+
+    # Save the record
+    @key_pair = KeyPair.new(
+        key_pair_params.merge(:user_name => @current_user.user_name,
```

Comment:
Indent the first parameter one step more than the previous line.<br>Use the new Ruby 1.9 hash syntax.